ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -549,6 +549,67 @@ public func concatMap<T, U, E>(transform: T -> SignalProducer<U, E>)(producer: S
     return producer |> map(transform) |> concat
 }
 
+/// Returns a signal that forwards values from the latest signal sent on
+/// `producer`, ignoring values sent on previous inner signals.
+///
+/// An error sent on `producer` or the latest inner signal will be sent on the
+/// returned signal.
+///
+/// The returned signal completes when `producer` and the latest inner signal
+/// have both completed.
+public func latest<T, E>(producer: SignalProducer<SignalProducer<T, E>, E>) -> SignalProducer<T, E> {
+	return SignalProducer<T, E> { sink, disposable in
+		let producerCompleted = Atomic(false)
+		let latestCompleted = Atomic(false)
+		
+		let completeIfNecessary: () -> () = {
+			if producerCompleted.value && latestCompleted.value {
+				sendCompleted(sink)
+			}
+		}
+
+		let latestDisposable = SerialDisposable()
+		disposable.addDisposable(latestDisposable)
+
+		producer.startWithSignal { signal, producerDisposable in
+			disposable.addDisposable(producerDisposable)
+			
+			signal.observe(
+				next: { signal in
+					latestCompleted.value = false
+					
+					signal.startWithSignal { signal, signalDisposable in
+						latestDisposable.innerDisposable = signalDisposable
+						
+						signal.observe(SinkOf { event in
+							switch event {
+							case .Completed:
+								latestCompleted.value = true
```

Comment:
Thanks. I saw that, but didn't pay close enough attention to see what it was saying.

> Setting `lastCompleted` needs to happen after disposal of the previous signal.

Is that sufficient to prevent a theoretical race? Even though the previous signal has been disposed, disposal doesn't interrupt any active observation callback. There's no _guarantee_ that the 1. completion event starts being handled, 2. it gets disposed, 3. `lastCompleted` set to false, 4. completion event sets `lastCompleted` to true.